### ISSUE TITLE
VDJServer's IgBlast

### ISSUE THREAD

**Issue by schristley** [airr-knowledge/issues#53]
- [x] port to tapis v3 app
- [x] utilize repertoire metadata to determine species, sequence type and germline (won't implement)
- [x] utilize AIRR germline sets
- [ ] #106
- [x] switch to igblast in repcalc container versus its own
- [x] switch parallel model, eliminate file splitting, use igblastn threads
- [x] Add C genes to generate `c_call`

**Comment by bcorrie** [comment]
@schristley a couple of questions about the igblast app.

I got the igblast job to run through Tapis, but it looks like there are a bunch of errors. Would you expect the Tapis3 version to work "out of the box"? If I do a make test-bcr the Tapis job runs successfully but the Tapis error log has a bunch of error messages for various steps in the processing.

I see that on the meta-api branch you are working on the igblast app as well. How do we want to handle that? I am continuing to work on the akc-apps branch at the moment.

I suppose there is a question as to whether I should continue to work on this (given the NIH thing) or not since I am just starting to ramp up on this Tapis App?

**Comment by schristley** [comment]
> [@schristley](https://github.com/schristley) a couple of questions about the igblast app.
> 
> I got the igblast job to run through Tapis, but it looks like there are a bunch of errors. Would you expect the Tapis3 version to work "out of the box"? If I do a make test-bcr the Tapis job runs successfully but the Tapis error log has a bunch of error messages for various steps in the processing.
> 
> I see that on the meta-api branch you are working on the igblast app as well. How do we want to handle that? I am continuing to work on the akc-apps branch at the moment.
> 
> I suppose there is a question as to whether I should continue to work on this (given the NIH thing) or not since I am just starting to ramp up on this Tapis App?

I have done a little work on it but yes it should mostly work "out of the box". I suspect the errors you are getting is the clonal analysis that runs after IgBlast, and maybe trying to assign repertoire_ids to the AIRR TSVs. It is ok to ignore that for now, so long as IgBlast is running properly and generating output files.

Don't worry too much about the Tapis-ization. Primarily I'd like to see if you can generate the BLAST files from the OGRDB germline set then modify the parameters when running IgBlast to use those files. That's like 90% of the work with the rest being how to package that up (which might be as simple as a TGZ file alongside the singularity images).

**Comment by schristley** [comment]
> [@schristley](https://github.com/schristley) a couple of questions about the igblast app.
> 
> I see that on the meta-api branch you are working on the igblast app as well. How do we want to handle that? I am continuing to work on the akc-apps branch at the moment.

The same changes should be on the master branch so maybe just pull those into your akc-apps branch.

**Comment by bcorrie** [comment]
> I have done a little work on it but yes it should mostly work "out of the box".

Have you been able to run it? I am getting the following for the paths for IGDATA:

`export IGDATA="/work/09939/bcorrie/ls6/../common/igblast-db/db.2019.01.23"`

and then:

`BLAST Database error: No alias or index file found for nucleotide database [/work/09939/bcorrie/ls6/../common/igblast-db/db.2019.01.23/germline//human/ReferenceDirectorySet/human_IG_V.fna] in search path [/scratch/09939/bcorrie/tapis/09d3a177-7822-4fba-a7af-17085dc41651-007/jobs/09d3a177-7822-4fba-a7af-17085dc41651-007`

and no AIRR tsv files being created.

Looks like something in the environment is messed up for the IG germline. Where is the germline DB supposed to be? Is it in the docker container?

**Comment by bcorrie** [comment]
Also, are the functions addOutputFile and addCalculation doing anything functional in the workflow? It looks to me like these are just capturing provenance information. Is that correct?

**Comment by schristley** [comment]
> > I have done a little work on it but yes it should mostly work "out of the box".
> 
> Have you been able to run it? I am getting the following for the paths for IGDATA:
> 
> `export IGDATA="/work/09939/bcorrie/ls6/../common/igblast-db/db.2019.01.23"`

Ah, dang, yeah I forgot about that. Those IgBlast files are in the work area, but vdj's work area, not yours.

> Looks like something in the environment is messed up for the IG germline. Where is the germline DB supposed to be? Is it in the docker container?

It's in the work area. Hmm so need to resolve that in some way. Maybe wrap that up in a TGZ and pull it into scratch instead of reference work.

**Comment by schristley** [comment]
> Also, are the functions addOutputFile and addCalculation doing anything functional in the workflow? It looks to me like these are just capturing provenance information. Is that correct?

Yes, primarily provenance, addOutputFile is used to tag the final output files for archiving, as we don't want to archive all of the temporary files. Though in test mode, you aren't archiving, so ...

**Comment by bcorrie** [comment]
> > Have you been able to run it? I am getting the following for the paths for IGDATA:
> > `export IGDATA="/work/09939/bcorrie/ls6/../common/igblast-db/db.2019.01.23"`
> 
> Ah, dang, yeah I forgot about that. Those IgBlast files are in the work area, but vdj's work area, not yours.

Can you tar them up for me and put them in tapis://data-storage.vdjserver.org/akc/germlines or something like that. I can then add them as an input file to the Job. By adding a new tarball I should be able to easily switch to an OGRDB germline set.

**Comment by schristley** [comment]
ok, there is a `db.2019.01.23.tgz` file in the akc/germlines directory.

**Comment by bcorrie** [comment]
Removed myself from the assignee for this task...

**Comment by schristley** [comment]
>  utilize repertoire metadata to determine species, sequence type and germline

thinking about this more, this may allow igblast to process both TCR and BCR in the same job but then presents difficulty with repcalc as it is designed to analyze just one locus at a time.

Instead, it's probably more appropriate to analyze the repertoires that have been selected, then set the locus parameter for the job. If the repertoires are both TCR and BCR then submit two jobs
